fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -138,100 +137,134 @@ public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull
-                            } else if (expression instanceof LuceneFieldKeyExpression) {
-                                Object value = entryKey.get(i + offset);
-                                if (value != null) {
-                                    insertDocumentField((LuceneFieldKeyExpression)expression, value, document, null);
-                                }
-                            }
-                        }
-                    }
-                } catch (IOException ioe) {
-                    throw new RecordCoreException("Issue updating index keys", "Key", entry, "IndexEntries", indexEntries,
-                            "savedRecord", savedRecord, ioe);
-                }
-            });
-            writer.addDocument(document);
-        } catch (Exception e) {
-            throw new RecordCoreException(e);
-        }
-        return AsyncUtil.DONE;
-    }
 
-    private void insertDocumentField(final LuceneFieldKeyExpression expression, Object value,
-                                     final Document document, String prefix) {
-        if (!(value == null && expression.getType() != LuceneKeyExpression.FieldType.STRING)) {
-            String fieldName = prefix == null ? expression.getFieldName() : prefix.concat("_").concat(expression.getFieldName());
-            switch (expression.getType()) {
-                case INT:
-                    // Todo: figure out how to expand functionality to include storage, sorting etc.
-                    document.add(new IntPoint(fieldName, (Integer)value));
-                    break;
-                case STRING:
-                    document.add(new TextField(fieldName, value == null ? "" : (String)value, expression.isStored() ? Field.Store.YES : Field.Store.NO));
-                    break;
-                case LONG:
-                    document.add(new LongPoint(fieldName, (long) value));
-                    break;
-                default:
-                    throw new RecordCoreArgumentException("Invalid type for lucene index field", "type", expression.getType());
+            // if old record has fields in the index and there are changes then delete old record
+            if (!oldRecordFields.isEmpty() && (!updatedFields.isEmpty() || newRecord == null)) {
+                Query query = SortedDocValuesField.newSlowExactQuery(PRIMARY_KEY_SEARCH_NAME, new BytesRef(oldRecord.getPrimaryKey().pack()));
+                writer.deleteDocuments(query);
             }
+
+            // if there are changes then update the index with the new document.
+            if (!updatedFields.isEmpty()) {
+                byte[] primaryKey = newRecord.getPrimaryKey().pack();
+                BytesRef ref = new BytesRef(primaryKey);
+                document.add(new StoredField(PRIMARY_KEY_FIELD_NAME, ref));
+                document.add(new SortedDocValuesField(PRIMARY_KEY_SEARCH_NAME, ref));
+                for (DocumentEntry entry : newRecordFields) {
```

Comment:
`newRecordFields` and `updatedFields` are the same, so this is only the changed fields. Don't we need them all?